### PR TITLE
Update design docs to reflect our new marshaller design from partner feedback

### PR DESCRIPTION
**Pull request by jkoritzinsky**
The interop team has been working on new designs for our source-generated marshallers. Now that we're starting implementation in the runtime of these new marshallers, let's make sure to follow our "write it down" culture so we have a record of why we did what we did for our future selves.

### CODE REVIEW

File: docs/design/libraries/LibraryImportGenerator/Pipeline.md [hunk 1]
```diff
@@ -97,9 +103,11 @@ try
     << Marshal >>
     << Pin >> (fixed)
     {
+        << Pinned Marshal >>
         << Invoke >>
     }
     << Keep Alive >>
```
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/LibraryImportGenerator/Pipeline.md:109]:
```suggestion
    << Notify For Successful Invoke >>
```

File: docs/design/libraries/LibraryImportGenerator/SpanMarshallers.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 As part of the exit criteria for the LibraryImportGenerator experiment, we have decided to introduce support for marshalling `System.Span<T>` and `System.ReadOnlySpan<T>` into the LibraryImportGenerator-generated stubs. This document describes design decisions made during the implementation of these marshallers.
 
+> NOTE: These design docs are kept for historical purposes. The designs in this file are supersceded by the designs in [UserTypeMarshallingV2.md](UserTypeMarshallingV2.md).
```
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/LibraryImportGenerator/SpanMarshallers.md:5]:
```suggestion
> NOTE: These design docs are kept for historical purposes. The designs in this file are superseded by the designs in [UserTypeMarshallingV2.md](UserTypeMarshallingV2.md).
```

File: docs/design/libraries/LibraryImportGenerator/StructMarshalling.md [hunk 3]
```diff
@@ -4,6 +4,8 @@ As part of the new source-generated direction for .NET Interop, we are looking a
 
 These types pose an interesting problem for a number of reasons listed below. With a few constraints, I believe we can create a system that will enable users to use their own user-defined types and pass them by-value to native code.
 
+> NOTE: These design docs are kept for historical purposes. The designs in this file are supersceded by the designs in [UserTypeMarshallingV2.md](UserTypeMarshallingV2.md).
```
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/LibraryImportGenerator/StructMarshalling.md:7]:
```suggestion
> NOTE: These design docs are kept for historical purposes. The designs in this file are superseded by the designs in [UserTypeMarshallingV2.md](UserTypeMarshallingV2.md).
```

**Review by AaronRobinsonMSFT — APPROVED** [review]
Let's make sure @elinor-fung signs off on this too.

File: docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+
+### Instance GetPinnableReference method on stateful shapes
+
+The non-static `GetPinnableReference` method on stateful shapes is provided to enable pinning a managed value as part of the marshalling process. As some types don't have values that need to be pinned to help with marshalling and pinning has some overhead, this member is optional to make the overhead pay-for-play.
+
+### Static GetPinnableReference method
+
+The static GetPinnableReference method provides a mechanism to pin a managed value and pass down the pinned value directly to native code. This allows us to provide massive performance benefits and to match built-in interop semantics. Unlike the previous design that used the `GetPinnableReference` method on the managed type in some scenarios, this design allows the "interop" pinning rules to not match the easier-to-use `GetPinnableReference` instance method, which may have differing semantics (`Span<T>` and arrays being a prime example here). As many types aren't marshallable via only pinning, the generator does not require this method on every marshaller.
+
+### `-Generated` method variants
```
Comment (elinor-fung) [inline docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md:717]: Can we add a section that gives the general picture for usage with `NativeMarshalling` / `MarshalUsing`? Those are only mentioned in the `StructMarshalling` doc, which has the note about being superseded by this one. But I feel like this one doesn't quite tie everything together, so you'd still need to read part of the other one and reason about usage.

File: docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+    private int hr;
+}
+
+[ManagedToUnmanagedMarshallers(typeof(HResult), InMarshaller = typeof(HResultMarshaller), RefMarshaller = typeof(HResultMarshaller), OutMarshaller = typeof(HResultMarshaller))]
+[UnmanagedToManagedMarshallers(typeof(HResult), InMarshaller = typeof(HResultMarshaller), RefMarshaller = typeof(HResultMarshaller), OutMarshaller = typeof(HResultMarshaller))]
+[ElementMarshaller(typeof(HResult), typeof(HResultMarshaller))]
+public static class HResultMarshaller
+{
+    public int ConvertToUnmanaged(HResult hr);
+    public HResult ConvertToManaged(int hr);
```
Comment (elinor-fung) [inline docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md:757]:
```suggestion
    public static int ConvertToUnmanaged(HResult hr);
    public static HResult ConvertToManaged(int hr);
```

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**